File: apps/api/src/application/products/delete-product.ts
import { NotFoundError } from "@/domain/shared/domain-error";

import type { ProductRepository } from "@/application/ports/product-repository";

export class DeleteProduct {
  constructor(private readonly products: ProductRepository) {}

  async execute(id: string): Promise<void> {
    const deleted = await this.products.softDelete(id);
    if (!deleted) {
      throw new NotFoundError("product", id);
    }
  }
}

File: apps/api/src/ecommerce_api/application/products/delete_product.py
from ecommerce_api.application.ports.product_repository import ProductRepository
from ecommerce_api.domain.errors import NotFoundError


class DeleteProduct:
    def __init__(self, products: ProductRepository) -> None:
        self._products = products

    async def execute(self, id: str) -> None:
        if not await self._products.soft_delete(id):
            raise NotFoundError('product', id)
